File: day4.py

from collections import namedtuple
import time
import os


def get_input():
    with open('input4') as fh:
        lines = list(fh)
        return map(lambda x: x.rstrip(), lines)

# [1518-11-01 00:00] Guard #10 begins shift
# [1518-11-01 00:05] falls asleep
# [1518-11-01 00:25] wakes up
def parse_line(line):
    def parse_id(act):
        act = act.replace(' #', '~')
        act = act.replace(' ', '~')
        spl = act.split('~')
        return int(spl[1])
    
    def parse_timestamp(timestamp):
        t = time_struct = time.strptime(timestamp, '%Y-%m-%d %H:%M')
        total_mins = t[4] + \
               t[3] * 60 + \
               t[2] * 24 * 60 + \
               t[1] * 30 * 24 * 60 + \
               t[0] * 365 * 24 * 60
        minute = t[4]
        date_string = str(t[1]) + '-' + str(t[2])
        return total_mins, minute, time_struct, date_string


    line = line.replace('[', '')
    line = line.replace('] ', '~')

    spl = line.split('~')
    ts = spl[0]
    act = spl[1]

    ret = {}
    ret['total_mins'], ret['min'], ret['ts'], ret['date_string'] = parse_timestamp(ts)
        
    if 'begins shift' in act:
        ret['id'] = parse_id(act)
        ret['act'] = 'begin'
    elif 'falls asleep' in act:
        ret['act'] = 'sleep'
    elif 'wakes up' in act:
        ret['act'] = 'wake'
    else:
        assert False, 'Failed to parse line: ' + line

    return ret

def sort_entries(data):
    def sort_cmp(a, b):
        return a['total_mins'] - b['total_mins']
    data.sort(cmp=sort_cmp)

def make_table(data):
    rows = []
    for d in data:
        print d['act'] + "\t" + d['date_string'] + "\tID: " + str(d.get('id', '')) + '\t' + str(d['min'])
        # new_row = {}
        # new_row['date'] = d['date_string']
        # new_row['id']

def go():
    input = get_input()
    data = []
    for line in input:
        data.append(parse_line(line))
    sort_entries(data)
    make_table(data)

go()

File: aocjava/src/Day4.java
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

enum Action{
    BEGIN,
    SLEEP,
    WAKE
}

class Line{
    public Line(){}
    public Line(Date date, Action action, int ID, int min){
        this.date = date;
        this.action = action;
        this.ID = ID;
        this.min = min;
    }
    public Date date;
    public Action action;
    public int ID;
    public int min;
}

class Guard{
    public int ID;
    public int maxSleepSize;
    public int maxSleepMinute;
}

class SleepResult{
    public SleepResult(){
        ID = 0;
        sleepMinsCount = 0;
        for (int i = 0; i < 60; i++) {
            sleepCount[i] = 0;
        }
    }
    public int ID;
    public int sleepMinsCount;
    public int[] sleepCount = new int[60];
}

public class Day4 {
    public static void main(String[] args){
        InputReader ir = new InputReader(4);
        ArrayList<Line> lines = new ArrayList<>();

        for(String line : ir.getLines()){
            lines.add(parseLine(line));
        }
        sortLines(lines);
        Map<Integer, SleepResult> sleepResultMap = calcAllSleepResults(lines);

        for (Map.Entry<Integer, SleepResult> entry : sleepResultMap.entrySet()){
            System.out.println("ID " + entry.getKey() + " Mins " + entry.getValue());
        }

        SleepResult result = findSleepiestGuard(sleepResultMap);

        System.out.println("[Part 1] Guard " + result.ID + " spends " + result.sleepMinsCount + " mins asleep.");
        int sleepiestMin = findSleepiestMinute(result);
        System.out.println("Sleeps the longest during minute #" + sleepiestMin);
        System.out.println("ID x min = " + result.ID * sleepiestMin);

        Guard g = findSleepiestMinuteGuard(sleepResultMap);
        System.out.println("[Part 2] Guard " + g.ID  + " spends " +
                g.maxSleepSize + " minutes asleep during minute #" +
                g.maxSleepMinute);
        System.out.println("ID x minute# = " + g.ID  * g.maxSleepMinute);
    }

    public static void sortLines(ArrayList<Line> lines){
        lines.sort((Line a, Line b) -> a.date.compareTo(b.date));
    }

    public static int findSleepiestMinute(SleepResult sleepResult){
        int maxIndex = 0;
        int maxValue = 0;
        for (int i = 0; i < sleepResult.sleepCount.length; i++) {
            if (sleepResult.sleepCount[i] > maxValue){
                maxIndex = i;
                maxValue = sleepResult.sleepCount[i];
            }
        }
        return maxIndex;
    }

    public static Guard findSleepiestMinuteGuard(Map<Integer, SleepResult> sleepResultsMap){
        Guard ret = new Guard();
        ret.ID = 0;
        ret.maxSleepSize = 0;
        ret.maxSleepMinute = 0;

        for(Map.Entry<Integer, SleepResult> entry : sleepResultsMap.entrySet()){
            int[] arr = entry.getValue().sleepCount;
            for (int i = 0; i < arr.length; i++) {
                if (arr[i] > ret.maxSleepSize){
                    ret.maxSleepMinute = i;
                    ret.maxSleepSize = arr[i];
                    ret.ID = entry.getKey();
                }
            }
        }
        return ret;
    }

    public static SleepResult findSleepiestGuard(Map<Integer, SleepResult> sleepResultsMap){
        int maxMinsAsleep = 0;
        int guardID = 0;
        for (Map.Entry<Integer, SleepResult> entry : sleepResultsMap.entrySet()) {
            if (entry.getValue().sleepMinsCount > maxMinsAsleep){
                guardID = entry.getKey();
                maxMinsAsleep = entry.getValue().sleepMinsCount;
            }
        }

        return sleepResultsMap.get(guardID);
    }


    public static SleepResult calcSleepResult(int startSleepMin, int endSleepMin, int ID, SleepResult previousResult){
        int timeSlept = endSleepMin - startSleepMin;

        previousResult.ID = ID;
        previousResult.sleepMinsCount += timeSlept;

        for (int i = startSleepMin; i < endSleepMin; i++) {
            previousResult.sleepCount[i]++;
        }

        return previousResult;
    }

    public static Map<Integer, SleepResult> calcAllSleepResults(ArrayList<Line> lines){
        int startSleepMin = 0;
        Map<Integer, SleepResult> ret = new HashMap<>();

        int curID = 0;
        for(Line l : lines){
            if (l.action == Action.BEGIN) {
                curID = l.ID;
            }

            else if (l.action == Action.SLEEP) {
                startSleepMin = l.min;
            }

            else if (l.action == Action.WAKE) {
                ret.put(curID, calcSleepResult(startSleepMin, l.min, curID,
                        ret.getOrDefault(curID, new SleepResult())));
            }
        }
        return ret;
    }

    public static Line parseLine(String line){
        //[1518-09-26 00:35] falls asleep
        //[1518-07-15 00:52] wakes up
        //[1518-06-08 00:36] wakes up
        //[1518-04-23 23:47]Guard #607 begins shift

        Line ret = new Line();

        if (line.contains("falls asleep"))
            ret.action = Action.SLEEP;
        else if (line.contains("wakes up"))
            ret.action = Action.WAKE;
        else if (line.contains("begins shift"))
            ret.action = Action.BEGIN;

        line = line.replace("[15", "20");
        line = line.replace("]", "~");
        line = line.replace(" #", "~");
        line = line.replace(" begins", "~");

        String[] tokens = line.split("~");
        try{
            ret.date = new SimpleDateFormat("yyyy-MM-dd hh:mm").parse(tokens[0]);
        }
        catch (ParseException ex){
            System.out.println("Parse exception!: " + ex);
        }

        if (ret.action == Action.BEGIN)
            ret.ID = Integer.parseInt(tokens[2]);

        ret.min = ret.date.getMinutes();
        System.out.println(ret.min);

        return ret;

    }
}
